src/plugins/topic_viewer/TopicViewer.cc

```diff
@@ -0,0 +1,330 @@
+  this->plotableTypes.push_back(FieldDescriptor::Type::TYPE_INT64);
+  this->plotableTypes.push_back(FieldDescriptor::Type::TYPE_UINT32);
+  this->plotableTypes.push_back(FieldDescriptor::Type::TYPE_UINT64);
+  this->plotableTypes.push_back(FieldDescriptor::Type::TYPE_BOOL);
+
+  this->plottingMode = false;
+
+  this->CreateModel();
+
+  ignition::gui::App()->Engine()->rootContext()->setContextProperty(
+                "TopicsModel", this->dataPtr->model);
+}
+
+//////////////////////////////////////////////////
+TopicViewer::~TopicViewer()
+{
+}
+
+//////////////////////////////////////////////////
+void TopicViewer::LoadConfig(const tinyxml2::XMLElement *)
+{
+  if (this->title.empty())
+    this->title = "Topic Viewer";
+}
+
+//////////////////////////////////////////////////
+void TopicViewer::CreateModel()
+{
+  this->dataPtr->model = new TopicsModel();
+
+  std::vector<std::string> topics;
+  this->dataPtr->node.TopicList(topics);
+
+  for (unsigned int i = 0; i < topics.size(); ++i)
+  {
+    std::vector<ignition::transport::MessagePublisher> infoMsgs;
+    this->dataPtr->node.TopicInfo(topics[i], infoMsgs);
+    std::string msgType = infoMsgs[0].MsgTypeName();
+    this->AddTopic(topics[i], msgType);
+  }
+}
+
+//////////////////////////////////////////////////
+void TopicViewer::SetPlottingMode(bool _mode)
+{
+  this->plottingMode = _mode;
+}
+
+//////////////////////////////////////////////////
+bool TopicViewer::PlottingMode()
+{
+  return this->plottingMode;
+}
+
+//////////////////////////////////////////////////
+void TopicViewer::AddTopic(const std::string &_topic,
+                           const std::string &_msg)
+{
+  // remove 'ignition.msgs.' from msg name
+  std::string msg = _msg;
+  try
+  {
+    if (msg.substr(0, 14) == "ignition.msgs.")
+      msg.erase(0, 14);
+  }
+  catch (std::out_of_range &exception)
+  {
+    igndbg << "faild to parse msg: " << _msg << std::endl;
+  }
+
+  QStandardItem *topicItem = this->FactoryItem(_topic, msg);
+  topicItem->setWhatsThis("Topic");
+  QStandardItem *parent = this->dataPtr->model->invisibleRootItem();
+  parent->appendRow(topicItem);
+
+  this->AddField(topicItem , msg, msg);
+}
+
+//////////////////////////////////////////////////
+void TopicViewer::AddField(QStandardItem *_parentItem,
+                           const std::string &_msgName,
+                           const std::string &_msgType)
+{
+  if (_msgName == "model")
+      return;
+
+  QStandardItem *msgItem;
+
+  // check if it is a topic, to skip the extra level of the topic Msg
+  if (_parentItem->whatsThis() == "Topic")
+  {
+    msgItem = _parentItem;
+    // make it different, so next iteration will make a new msg item
+    msgItem->setWhatsThis("Msg");
+  }
+  else
+  {
+    msgItem = this->FactoryItem(_msgName, _msgType);
+    _parentItem->appendRow(msgItem);
+  }
+
+  auto msg = ignition::msgs::Factory::New(_msgType);
+  if (!msg)
+      return;
```

Comment:
Do we want to print a warning if we can't create a message?

Reply:
there is already a lot of msgs types that will be like that like [Map] for example or any msg that is not ign-msg